src/crypto/PublicKey.ts

```diff
@@ -211,6 +217,123 @@ export class SignedPublicKey
   }
 }
 
+export class AccountLinkedPublicKeyV1
+  extends UnsignedPublicKey
+  implements publicKey.AccountLinkedPublicKey_V1
+{
+  keyBytes: Uint8Array
+  staticSignature: AccountLinkedStaticSignature | undefined
+  siweSignature: signature.AccountLinkedSIWESignature | undefined
+
+  constructor(obj: publicKey.AccountLinkedPublicKey_V1) {
+    if (!obj.keyBytes) {
+      throw new Error('Invalid AccountLinkedPublicKeyV1 has no keyBytes')
+    }
+    super(publicKey.UnsignedPublicKey.decode(obj.keyBytes))
+    this.keyBytes = obj.keyBytes
+    if (obj.staticSignature) {
+      this.staticSignature = new AccountLinkedStaticSignature(
+        obj.staticSignature
+      )
+    } else {
+      throw new Error(
+        'Unimplemented signature type for AccountLinkedPublicKeyV1'
+      )
+    }
+  }
+
+  // Return bytes of the encoded unsigned key.
+  private bytesToSign(): Uint8Array {
+    return this.keyBytes
+  }
+
+  public getLinkedAddress(role: AccountLinkedRole): string {
+    if (this.staticSignature) {
+      // Decode from UTF-8
+      const signatureText = new TextDecoder().decode(this.staticSignature.text)
+      if (
+        signatureText !==
+        WalletSigner.accountLinkRequestText(this.bytesToSign(), role)
+      ) {
+        throw new Error('Signature text does not match expected text')
+      }
+      const digest = hexToBytes(utils.hashMessage(signatureText))
+      const signature = this.staticSignature.walletEcdsaCompact
+      const publicKey = ecdsaSignerKey(digest, signature)
+      if (!publicKey) {
+        throw new Error('Static signature is not valid for given role')
```

Comment:
I'm not sure if this error would get thrown for an invalid signature, I think it must just return an unexpected ethereum address

Reply:
I thought I remembered from mucking around with tests a while back, for example with this code change:
```
diff --git a/src/crypto/Signature.ts b/src/crypto/Signature.ts
index 808838b..e091231 100644
--- a/src/crypto/Signature.ts
+++ b/src/crypto/Signature.ts
@@ -36,6 +36,7 @@ export function ecdsaSignerKey(
   digest: Uint8Array,
   signature: ECDSACompactWithRecovery
 ): UnsignedPublicKey | undefined {
+  digest[0] &= 0x1f // clear recovery bit
   const bytes = secp.recoverPublicKey(
     digest,
     signature.bytes,
```

you'll see failures like:

```
  ● InMemoryKeystore › getAccountAddress › returns the wallet address

    expect(received).toEqual(expected) // deep equality

    Expected: "0xacc9B903129E2693b0c09720cBa3972e35666066"
    Received: "0x3Db7e37217f83a89A222Ff7817a957ec393D1bA1"

      516 |       const returnedAddress = await aliceKeystore.getAccountAddress()
      517 |
    > 518 |       expect(aliceAddress).toEqual(returnedAddress)
          |                            ^
      519 |     })
      520 |   })
      521 | })

      at test/keystore/InMemoryKeystore.test.ts:518:28
      at fulfilled (test/keystore/InMemoryKeystore.test.ts:5:58)
```

Reply:
Yeah the way the error was worded made me a bit worried we were counting on the check, even the check now still seems a bit meaningless because it will only ever trigger in exceptionally rare circumstances (I think the r,s values have to hit an edge case in the math).

As a reader/consumer of the API, the `if` statement and error appears to catch "bad" signatures but it'll let basically any signature through. I know this seems like a nitpick but the first time I read it I genuinely though it was meant to be a signature validation check.

One solution: for an `AccountLinkedPublicKey` why not just carry the expected address as part of the object? Semantically it makes sense that an AccountLinkedPublicKey should have an expected account linked with it.  Then we could actually do a full validity check here (recovery key from signature, compare it).